### PR TITLE
uptrackmanager reuse

### PR DESCRIPTION
**Pull request by boks1971**

### CODE REVIEW

File: pkg/rtc/mediatrack.go
Comment (cnderrauber) [inline pkg/rtc/mediatrack.go:52]: why not continue to use sfu.Receiver, need to cast to sfu.Receiver in some place

**Review by cnderrauber — APPROVED** [review]
Nice work! just question for rename

File: pkg/rtc/mediatrack.go
Comment (boks1971) [reply pkg/rtc/mediatrack.go:52]: Thank you @cnderrauber . I want to reuse the receiver parts from `MediaTrack` also with the remote media track. So, using the interface and casting as necessary for now. That will allow to sharpen the interfaces further and make re-use better.

This PR will have conflicts with your session migration stuff @cnderrauber . So, was waiting for you to merge and then resolve conflicts and ask for review. But, you already did it :-) But, I will wait for you to merge, resolve conflicts and take it from there.

This allows full re-use of uptrackmanager so that remote participant becomes a lot smaller.

More to come...

@davidzhao Might as well ask you for review also now as Jie has already looked at it :-). But, will resolve conflicts after session migration and then only merge this.

File: pkg/rtc/localparticipant.go
Comment (boks1971) [inline pkg/rtc/localparticipant.go:1]: `LocalParticipant` split out to handle all the pending tracks bits. This is where there will be a bunch of conflicts with session migration PR which I will resolve.

Ideally, we will make an interface of this and make a NoOp impl and use it in cloud

File: pkg/rtc/mediatracksubscriptions.go
Comment (boks1971) [inline pkg/rtc/mediatracksubscriptions.go:258]: Doing this here while a subscriber is removed because we start a goroutine to close down track and were removing this subscription in the down track close callback.

But, when adding subscriber we were also checking if it is already subscribed. In the edge case of unsubscription followed by a subscription very close to each other, we may not have gotten the down track close callback and we will not process the subscription as already subscribed check will ignore the second subscription.

File: pkg/rtc/mediatrack.go
Comment (cnderrauber) [reply pkg/rtc/mediatrack.go:52]: Sounds good, Thanks!

File: pkg/rtc/localparticipant.go
Comment (boks1971) [inline pkg/rtc/localparticipant.go:179]: @cnderrauber This was in `UptrackManager`. It was moved here.

You were passing in `*ParticipantImpl` and doing the mid in this function. I did not want to pass in the parent pointer. So, calculating the `mid` at the call site and passing it in.

Also, this module does not have `onTrackUpdated` callback. So, returning indication if track is new and the call site fires `onTrackUpdated` if necessary.

File: pkg/rtc/mediatrackreceiver.go
Comment (boks1971) [inline pkg/rtc/mediatrackreceiver.go:1]: The next split out module to be able re-use in cloud.

File: pkg/rtc/room.go
Comment (boks1971) [inline pkg/rtc/room.go:638]: This lock looked redundant. So, removed it.

File: pkg/rtc/types/interfaces.go
Comment (boks1971) [inline pkg/rtc/types/interfaces.go:173]: Cleaning up interfaces more.

File: pkg/sfu/receiver.go
Comment (boks1971) [inline pkg/sfu/receiver.go:198]: Routing `onAudioLevel` via `sfu.Receiver` so that we can add it `sfu.TrackReceiver` interface and use it across cloud also.

**Comment by boks1971** [comment]
@cnderrauber @davidzhao This is ready for review again. Have merged the session migration PR (Jie, will really appreciate if you can use this branch to do a quick check that migration still works).

Also, split out `MediaTrackReceiver`. Will be submitting a cloud PR with some significant chunks of code removed :-)

File: pkg/rtc/localparticipant.go
Comment (davidzhao) [inline pkg/rtc/localparticipant.go:165]: is DTX only for the first audio track? or should this check cover if any audio tracks have DTX enabled?

File: pkg/rtc/room.go
Comment (davidzhao) [reply pkg/rtc/room.go:638]: this was added to prevent two goroutines from acquiring the exact timestamp/updates. If multiple requests come in to broadcastParticipantState, we would end up having a race that end up "erasing" one of the published tracks.

Since Go generates monotonic timestamps, the idea is to ensure the order of the content matches the ordering of Time. i.e. if a timestamp is later, then the accompanying `updates` is also more recent.

there may be better ways of preventing this race condition.

File: pkg/sfu/receiver.go
Comment (davidzhao) [reply pkg/sfu/receiver.go:198]: With speaker updates, I know we said that each SFU could perform local computation for remote tracks. I don't think we need to do that anymore. SpeakerUpdates are being forwarded to all media nodes.

**Review by davidzhao — APPROVED** [review]
really solid cleanup!

File: pkg/rtc/room.go
Comment (boks1971) [reply pkg/rtc/room.go:638]: Ouch, I thought it was locking some state and was forgotten when some code was removed. Will add it back and add a comment around it too.

File: pkg/sfu/receiver.go
Comment (boks1971) [reply pkg/sfu/receiver.go:198]: Yeah, that's true. Okay, let me move all the audio level stuff back into `mediatrack.go`

File: pkg/rtc/localparticipant.go
Comment (boks1971) [reply pkg/rtc/localparticipant.go:165]: okay, will change it to look at all. We ideally want per track setting, but this is the edge case. So, looking through all makes sense.

**Comment by boks1971** [comment]
@davidzhao Made this commit to address your comments - https://github.com/livekit/livekit-server/pull/318/commits/4457fe2e664dfb7272c766944a3ed7a3bc5bbbdf

**Comment by davidzhao** [comment]
> @davidzhao Made this commit to address your comments - [4457fe2](https://github.com/livekit/livekit-server/commit/4457fe2e664dfb7272c766944a3ed7a3bc5bbbdf)

Great! Thanks for adding comments around it.

**Pull request merged**
